File: agent/app/src/main/java/com/adbtoolkit/agent/server/ApiRouter.kt
package com.adbtoolkit.agent.server

import android.content.Context
import android.util.Log
import com.adbtoolkit.agent.api.*
import com.adbtoolkit.agent.security.PairingManager
import fi.iki.elonen.NanoHTTPD
import fi.iki.elonen.NanoHTTPD.Response

/**
 * Routes incoming HTTP requests to the appropriate API handler.
 *
 * URL scheme: `/api/<domain>/<action>[/<param>]`
 *
 * Security layers:
 * - Local/ADB requests: authenticated via X-Agent-Token
 * - Peer-to-peer requests: authenticated via X-Peer-Id + X-Peer-Signature (HMAC)
 * - Pairing endpoints: open (they ARE the authentication step)
 */
class ApiRouter(context: Context) {

    companion object {
        private const val TAG = "ApiRouter"
    }

    private val fileApi = FileApi(context)
    private val appApi = AppApi(context)
    private val contactsApi = ContactsApi(context)
    private val smsApi = SmsApi(context)
    private val deviceApi = DeviceApi(context)
    private val shellApi = ShellApi(context)
    private val pythonApi = PythonApi(context)
    private val peerApi = PeerApi(context)
    private val orchestratorApi = OrchestratorApi(context)
    private val pairingManager = PairingManager.getInstance(context)

    fun route(
        method: NanoHTTPD.Method,
        uri: String,
        session: NanoHTTPD.IHTTPSession,
    ): Response {
        // Normalise path
        val path = uri.trimEnd('/')
        val parts = path.removePrefix("/api/").split("/", limit = 3)
        val domain = parts.getOrNull(0)

        // ── Peer-authenticated requests ──────────────────────────
        // If request carries X-Peer-Id header, validate HMAC before
        // allowing access to data APIs (but NOT pairing endpoints).
        val peerId = session.headers["x-peer-id"]
        if (peerId != null && domain !in listOf("ping", "peer")) {
            val validation = pairingManager.validatePeerRequest(
                method = method.name,
                uri = uri,
                headers = session.headers,
            )
            if (!validation.valid) {
                Log.w(TAG, "Peer auth rejected: ${validation.reason}")
                return AgentServer.jsonError(
                    "Autenticação P2P falhou: ${validation.reason}",
                    Response.Status.FORBIDDEN
                )
            }
            Log.d(TAG, "Peer auth OK: $peerId → $path")
        }

        return when (domain) {
            "ping"         -> AgentServer.jsonOk(mapOf(
                "status" to "alive",
                "version" to "1.0.0",
                "python_ready" to pythonApi.isReady(),
                "device_id" to pairingManager.deviceId,
                "paired_count" to pairingManager.getPairedDevices().size,
            ))

            "files"        -> fileApi.handle(method, parts.drop(1), session)
            "apps"         -> appApi.handle(method, parts.drop(1), session)
            "contacts"     -> contactsApi.handle(method, parts.drop(1), session)
            "sms"          -> smsApi.handle(method, parts.drop(1), session)
            "device"       -> deviceApi.handle(method, parts.drop(1), session)
            "shell"        -> shellApi.handle(method, parts.drop(1), session)
            "python"       -> pythonApi.handle(method, parts.drop(1), session)
            "peer"         -> peerApi.handle(method, parts.drop(1), session)
            "orchestrator" -> orchestratorApi.handle(method, parts.drop(1), session)

            else -> AgentServer.jsonError("Unknown endpoint: $path",
                Response.Status.NOT_FOUND)
        }
    }
}

File: agent-ios/AgentIOS/Sources/Server/ApiRouter.swift
import Foundation

// ─────────────────────────────────────────────────────────────────────
//  ApiRouter.swift — Routes HTTP requests to API handlers
//
//  Same URL scheme as Android: /api/<domain>/<action>[/<param>]
// ─────────────────────────────────────────────────────────────────────

final class ApiRouter {

    private let contactsApi = ContactsHandler()
    private let photosApi = PhotosHandler()
    private let filesApi = FilesHandler()
    private let deviceApi = DeviceHandler()
    private let peerApi: PeerHandler
    private let pairingManager: PairingManager

    init(pairingManager: PairingManager) {
        self.pairingManager = pairingManager
        self.peerApi = PeerHandler(pairingManager: pairingManager)
    }

    func route(request: HTTPRequest) -> HTTPResponse {
        let path = request.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let parts = path
            .replacingOccurrences(of: "api/", with: "", options: .anchored)
            .components(separatedBy: "/")

        let domain = parts.first ?? ""
        let subParts = Array(parts.dropFirst())

        // P2P auth check (if X-Peer-Id header present)
        let peerId = request.headers["x-peer-id"]
        if let peerId = peerId, domain != "ping" && domain != "peer" {
            let isValid = pairingManager.validatePeerRequest(
                method: request.method,
                uri: request.path,
                headers: request.headers
            )
            if !isValid {
                AgentState.shared.log("router", "Peer auth rejected for \(peerId)")
                return .error("P2P authentication failed", status: .forbidden)
            }
        }

        switch domain {
        case "ping":
            return .ok([
                "status": "alive",
                "version": AgentConstants.appVersion,
                "platform": "ios",
                "device_id": pairingManager.deviceId,
                "paired_count": pairingManager.pairedDeviceCount,
            ])

        case "contacts":
            return contactsApi.handle(method: request.method, parts: subParts, request: request)

        case "photos":
            return photosApi.handle(method: request.method, parts: subParts, request: request)

        case "files":
            return filesApi.handle(method: request.method, parts: subParts, request: request)

        case "device":
            return deviceApi.handle(method: request.method, parts: subParts, request: request)

        case "peer":
            return peerApi.handle(method: request.method, parts: subParts, request: request)

        default:
            return .error("Unknown endpoint: /api/\(domain)", status: .notFound)
        }
    }
}
